util/http/http.go

```diff
@@ -2,27 +2,104 @@ package http
+const maxLength = 4093
+
 // MakeCookieMetadata generates a string representing a Web cookie.  Yum!
-func MakeCookieMetadata(key, value string, flags ...string) (string, error) {
-	components := []string{
-		fmt.Sprintf("%s=%s", key, value),
+func MakeCookieMetadata(key, value string, flags ...string) ([]string, error) {
+	attributes := strings.Join(flags, "; ")
+
+	// cookie: name=value; attributes and key: key-(i) e.g. argocd.token-0
+	maxValueLength := maxValueLength(key, attributes)
+	numberOfCookies := int(math.Ceil(float64(len(value)) / float64(maxValueLength)))
+	if numberOfCookies > maxNumber {
+		return nil, fmt.Errorf("invalid cookie value, at %d long it is longer than the max length of %d", len(value), maxValueLength*maxNumber)
+	}
+
+	return splitCookie(key, value, attributes), nil
+}
+
+// browser has limit on size of cookie, currently 4kb. In order to
+// support cookies longer than 4kb, we split cookie into multiple 4kb chunks.
+func splitCookie(key, value, attributes string) []string {
+	var cookies []string
+	valueLength := len(value)
+
+	numberOfChunks := int(math.Ceil(float64(valueLength) / float64(maxValueLength(key, attributes))))
+
+	// cookie: name=value; attributes and key: key-(i) e.g. argocd.token-0
+	maxValueLength := maxValueLength(key, attributes)
+	var end int
+	for i, j := 0, 0; i < valueLength; i, j = i+maxValueLength, j+1 {
+		end = i + maxValueLength
+		if end > valueLength {
+			end = valueLength
+		}
+
+		// first chunk will be of format argocd.token=<numberOfChunks>:token; attributes
+		if j == 0 && attributes == "" {
+			cookies = append(cookies, fmt.Sprintf("%s=%d:%s", key, numberOfChunks, value[i:end]))
+		} else if j == 0 && attributes != "" {
+			cookies = append(cookies, fmt.Sprintf("%s=%d:%s; %s", key, numberOfChunks, value[i:end], attributes))
+		} else if attributes == "" {
+			cookies = append(cookies, fmt.Sprintf("%s-%d=%s", key, j, value[i:end]))
+		} else {
+			cookies = append(cookies, fmt.Sprintf("%s-%d=%s; %s", key, j, value[i:end], attributes))
+		}
 	}
-	components = append(components, flags...)
-	header := strings.Join(components, "; ")
+	return cookies
+}
+
+// JoinCookies combines chunks of cookie based on key as prefix. It returns cookie
+// value as string. cookieString is of format key1=value1; key2=value2; key3=value3
+// first chunk will be of format argocd.token=<numberOfChunks>:token; attributes
+func JoinCookies(key string, cookieString string) (string, error) {
+	cookies := make(map[string]string)
+	for _, cookie := range strings.Split(cookieString, ";") {
+		parts := strings.Split(cookie, "=")
+		if len(parts) < 2 {
+			continue
+		}
+		cookies[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
+	}
+
+	var sb strings.Builder
+	firstPart, ok := cookies[key]
+	if !ok {
+		return "", fmt.Errorf("failed to retrieve cookie %s", key)
+	}
+	token := strings.Split(firstPart, ":")
+	if len(token) < 2 {
+		return "", fmt.Errorf("invalid cookie for key %s", key)
```

Comment:
For backward compatibility (to support users who manually set cookie in scripts  ) we need to assume one chunk, if number of chunks in not specified in first chunk.